File: Go/internal/net/transmitter.go
package net

import (
	"encoding/binary"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/GHeadFirst/netze-project/Go/internal/udp_packets"
)

type TransmissionStats struct {
	Timestamp    string  `json:"timestamp"`
	File         string  `json:"file"`
	FileSize     int64   `json:"file_size"`
	Duration     float64 `json:"duration"`
	Throughput   float64 `json:"throughput"`
	TotalPackets int     `json:"total_packets"`
	AcksReceived int     `json:"acks_received"`
	Retries      int     `json:"retries"`
	SuccessRate  float64 `json:"success_rate"`
}

func Transmitter(filename string) {
	const numTransmissions = 10
	const maxRetries = 3
	const serverAddr = "0.0.0.0:4010" // Match Python receiver's address

	// Create results directory if it doesn't exist
	resultsDir := "results"
	if err := os.MkdirAll(resultsDir, 0755); err != nil {
		log.Fatal("Error creating results directory:", err)
	}

	// Open log file
	logFile, err := os.OpenFile(filepath.Join(resultsDir, "results-go-transmitter.txt"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Fatal("Error opening log file:", err)
	}
	defer logFile.Close()

	fmt.Printf("\nStarting %d transmissions of %s\n", numTransmissions, filename)
	fmt.Println(strings.Repeat("=", 50))

	successfulTransmissions := 0
	for i := 0; i < numTransmissions; i++ {
		fmt.Printf("\nStarting transmission %d/%d\n", i+1, numTransmissions)

		// Get file info
		fileInfo, err := os.Stat(filename)
		if err != nil {
			log.Fatalf("Error getting file info: %v", err)
		}
		fileSize := fileInfo.Size()
		fmt.Printf("File size: %d bytes\n", fileSize)

		// Open file
		file, err := os.Open(filename)
		if err != nil {
			log.Fatalf("Error opening file: %v", err)
		}
		defer file.Close()

		// Create UDP connection
		conn, err := net.Dial("udp", serverAddr)
		if err != nil {
			log.Fatalf("Error creating UDP connection: %v", err)
		}
		defer conn.Close()

		// Set read timeout
		conn.SetReadDeadline(time.Now().Add(5 * time.Second))

		// Generate random sequence ID
		sequenceID := uint16(rand.Intn(65536))
		fmt.Printf("Starting transmission with sequence ID: %d\n", sequenceID)

		// Calculate total number of packets
		totalPackets := (fileSize + 1023) / 1024 // Round up division with 1023 because Go rounds down normally
		fmt.Printf("Total packets to send: %d\n", totalPackets)

		// Send first packet
		firstPacket := udp_packets.First_packet{
			Head: udp_packets.Header{
				Transmission_id: sequenceID,
				Sequence_number: 0,
			},
			Max_sequence_number: uint32(totalPackets),
			File_Name:           filename,
		}
		firstPacketBytes := BuildPacket(&firstPacket, [16]byte{})
		_, err = conn.Write(firstPacketBytes)
		if err != nil {
			log.Fatalf("Error sending first packet: %v", err)
		}

		// Send file in chunks
		buffer := make([]byte, 1024)
		sequenceNumber := uint32(1) // Start from 1 since 0 is first packet
		bytesSent := int64(0)
		startTime := time.Now()
		acksReceived := 0
		retries := 0
		success := true

		for {
			n, err := file.Read(buffer)
			if err == io.EOF {
				break
			}
			if err != nil {
				log.Fatalf("Error reading file: %v", err)
			}

			// Create and send data packet
			dataPacket := udp_packets.Data_packet{
				Head: udp_packets.Header{
					Transmission_id: sequenceID,
					Sequence_number: sequenceNumber,
				},
				Data: buffer[:n],
			}

			packetBytes := BuildPacket(&dataPacket, [16]byte{})
			retryCount := 0
			for retryCount < maxRetries {
				_, err = conn.Write(packetBytes)
				if err != nil {
					log.Fatalf("Error sending packet: %v", err)
				}

				// Wait for ACK
				ackBuffer := make([]byte, 3) // ACK is 3 bytes ('ACK')
				_, err = conn.Read(ackBuffer)
				if err != nil {
					if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
						fmt.Printf("\nTimeout waiting for ACK for packet %d, retry %d/%d\n",
							sequenceNumber, retryCount+1, maxRetries)
						retryCount++
						retries++
						continue
					}
					log.Fatalf("Error receiving ACK: %v", err)
				}

				// Check if ACK is valid
				if string(ackBuffer) != "ACK" {
					fmt.Printf("\nUnexpected ACK received: %s, retry %d/%d\n",
						string(ackBuffer), retryCount+1, maxRetries)
					retryCount++
					retries++
					continue
				}

				acksReceived++
				break
			}

			if retryCount >= maxRetries {
				fmt.Printf("\nMax retries reached for packet %d, aborting transmission\n", sequenceNumber)
				success = false
				break
			}

			bytesSent += int64(n)
			progress := float64(bytesSent) / float64(fileSize) * 100
			elapsed := time.Since(startTime)
			speed := float64(bytesSent) / elapsed.Seconds() / 1024 / 1024 // MB/s

			fmt.Printf("\rProgress: %.2f%% (%d/%d packets) - Speed: %.2f MB/s",
				progress, sequenceNumber, totalPackets, speed)

			sequenceNumber++
		}

		if success {
			// Calculate MD5
			md5Sum := CalcMD5(filename)

			// Send last packet
			lastPacket := udp_packets.Last_packet{
				Head: udp_packets.Header{
					Transmission_id: sequenceID,
					Sequence_number: sequenceNumber,
				},
				MD5: md5Sum,
			}
			lastPacketBytes := BuildPacket(&lastPacket, md5Sum)
			_, err = conn.Write(lastPacketBytes)
			if err != nil {
				log.Fatalf("Error sending last packet: %v", err)
			}

			totalTime := time.Since(startTime)
			fmt.Printf("\n\nTransfer completed!\n")
			fmt.Printf("Total time: %.2f seconds\n", totalTime.Seconds())
			fmt.Printf("Average speed: %.2f MB/s\n", float64(fileSize)/totalTime.Seconds()/1024/1024)
			fmt.Printf("Total packets sent: %d\n", sequenceNumber+1)

			// Log statistics
			stats := TransmissionStats{
				Timestamp:    time.Now().Format(time.RFC3339),
				File:         filename,
				FileSize:     fileSize,
				Duration:     totalTime.Seconds(),
				Throughput:   float64(fileSize) / totalTime.Seconds(),
				TotalPackets: int(sequenceNumber + 1),
				AcksReceived: acksReceived,
				Retries:      retries,
				SuccessRate:  float64(acksReceived) / float64(sequenceNumber+1) * 100,
			}

			// Write to log file
			logEntry := fmt.Sprintf("\n%s\n", strings.Repeat("=", 50))
			logEntry += fmt.Sprintf("Transmission %d/%d at %s\n", i+1, numTransmissions, stats.Timestamp)
			logEntry += fmt.Sprintf("File: %s\n", stats.File)
			logEntry += fmt.Sprintf("File Size: %.2f KB\n", float64(stats.FileSize)/1024)
			logEntry += fmt.Sprintf("Duration: %.2f seconds\n", stats.Duration)
			logEntry += fmt.Sprintf("Throughput: %.2f KB/s\n", stats.Throughput/1024)
			logEntry += fmt.Sprintf("Total Packets: %d\n", stats.TotalPackets)
			logEntry += fmt.Sprintf("ACKs Received: %d\n", stats.AcksReceived)
			logEntry += fmt.Sprintf("Retries: %d\n", stats.Retries)
			logEntry += fmt.Sprintf("Success Rate: %.1f%%\n", stats.SuccessRate)
			logEntry += fmt.Sprintf("%s\n", strings.Repeat("=", 50))

			if _, err := logFile.WriteString(logEntry); err != nil {
				log.Fatal("Error writing to log file:", err)
			}

			successfulTransmissions++
		}

		fmt.Printf("\nCompleted transmission %d/%d\n", i+1, numTransmissions)
		fmt.Println(strings.Repeat("-", 50))
	}

	// Print summary
	fmt.Printf("\nTransmission Summary:\n")
	fmt.Printf("Total transmissions: %d\n", numTransmissions)
	fmt.Printf("Successful: %d\n", successfulTransmissions)
	fmt.Printf("Failed: %d\n", numTransmissions-successfulTransmissions)
	fmt.Printf("Success rate: %.1f%%\n", float64(successfulTransmissions)/float64(numTransmissions)*100)
	fmt.Printf("Results have been saved to results/results-go-transmitter.txt\n")
}

func BuildPacket(x udp_packets.Packet, md5_byte [16]byte) []byte {
	var packet []byte

	switch pkt := x.(type) {

	case *udp_packets.First_packet:
		var transmission_byte [2]byte
		var sequence_byte [4]byte
		var max_sequence_byte [4]byte
		var file_name_byte [256]byte

		binary.BigEndian.PutUint16(transmission_byte[:], uint16(pkt.Head.Transmission_id))
		binary.BigEndian.PutUint32(sequence_byte[:], pkt.Head.Sequence_number)
		binary.BigEndian.PutUint32(max_sequence_byte[:], pkt.Max_sequence_number)
		copy(file_name_byte[:], []byte(pkt.File_Name))

		packet = append(packet, transmission_byte[:]...)
		packet = append(packet, sequence_byte[:]...)
		packet = append(packet, max_sequence_byte[:]...)
		packet = append(packet, file_name_byte[:]...)

	case *udp_packets.Data_packet:
		var transmission_byte [2]byte
		var sequence_byte [4]byte
		// data is already in bytes

		binary.BigEndian.PutUint16(transmission_byte[:], uint16(pkt.Head.Transmission_id))
		binary.BigEndian.PutUint32(sequence_byte[:], pkt.Head.Sequence_number)

		packet = append(packet, transmission_byte[:]...)
		packet = append(packet, sequence_byte[:]...)
		packet = append(packet, pkt.Data...)

	case *udp_packets.Last_packet:
		var transmission_byte [2]byte
		var sequence_byte [4]byte
		//md5 is already in [16]byte

		binary.BigEndian.PutUint16(transmission_byte[:], uint16(pkt.Head.Transmission_id))
		binary.BigEndian.PutUint32(sequence_byte[:], pkt.Head.Sequence_number)

		packet = append(packet, transmission_byte[:]...)
		packet = append(packet, sequence_byte[:]...)
		packet = append(packet, md5_byte[:]...)
	}

	return packet
}

func create_header(id uint16, sequence uint32) udp_packets.Header {
	return udp_packets.Header{Transmission_id: id, Sequence_number: sequence}
}

func create_first_packet(head udp_packets.Header, max_sequence_number uint32, filename string) udp_packets.First_packet {
	return udp_packets.First_packet{Head: head, Max_sequence_number: max_sequence_number, File_Name: filename}
}

func create_data_packet(head udp_packets.Header, buf []byte) udp_packets.Data_packet {
	return udp_packets.Data_packet{Head: head, Data: buf}
}

func create_last_packet(head udp_packets.Header, md5_byte [16]byte) udp_packets.Last_packet {
	return udp_packets.Last_packet{Head: head, MD5: md5_byte}
}

File: Python/transmitter.py
import socket
from packets import *
import math
from pathlib import Path
from hashlib import md5
import time
import sys
from datetime import datetime


# file_name = test_data.txt
project_root = Path(__file__).resolve().parent.parent
transmission_id = 0
max_sequence_number = None

class FileReader():

    def __init__(self,file_name) -> None:
        file_to_read = project_root / file_name
        self.file = file_to_read
        self.file_name = file_to_read.name
        try:
            file_size = file_to_read.stat().st_size
            self.file_size = file_size
            with file_to_read.open("rb") as f:
                contents = f.read()
                self.contents = contents
                self.md5_hash = md5(self.contents).digest() 
            
        except FileNotFoundError:
            raise FileNotFoundError(f"File {file_name} not found.")

    def get_chunk(self,offset,amount_of_bytes) -> bytes:
        if offset >= self.file_size:
            raise ValueError(f"Offset {offset} exceeds file size {self.file_size}")

        if amount_of_bytes <= 0:
           raise ValueError("amount_of_bytes must be positive")

        chunk_of_data = self.contents[offset: offset + amount_of_bytes]
        return chunk_of_data
    
    def get_total_chunks(self,buffer_size) -> int:
        return math.ceil(self.file_size / buffer_size)


class PacketBuilder():
    def __init__(self,file_name,buffer_size,transmission_id) -> None:
        # Our class Variables
        self.sequence_number = 0
        self.transmission_id = transmission_id

        file = FileReader(file_name)
        self.file = file
        self.buffer_size = buffer_size
        self.first_packet = None
        self.data_packets = []
        self.last_packet = None

    def create_first_packet(self) -> FirstPacket:
        self.max_sequence_number = self.file.get_total_chunks(self.buffer_size)
        self.first_packet = FirstPacket(self.transmission_id, self.sequence_number,
            self.max_sequence_number, self.file.file_name)
        self.sequence_number += 1

    def create_data_packet(self) -> DataPacket:
        # Create data packets from 1 to max_sequence_number
        for i in range(1, self.max_sequence_number + 1):
            offset = (i - 1) * self.buffer_size
            chunk = self.file.get_chunk(offset, self.buffer_size)
            new_data_packet = DataPacket(self.transmission_id, i, chunk)
            self.data_packets.append(new_data_packet)

    def create_last_packet(self) -> LastPacket:
        # Last packet should be at max_sequence_number + 1
        self.last_packet = LastPacket(self.transmission_id, self.max_sequence_number + 1, self.file.md5_hash) 

    def get_all_packets(self) -> list:
        self.create_first_packet()
        self.create_data_packet()
        self.create_last_packet()
        return [self.first_packet] + self.data_packets + [self.last_packet]

class TransmitterLogger:
    def __init__(self):
        self.results_dir = Path("results")
        self.results_dir.mkdir(exist_ok=True)
        self.timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        self.log_file = self.results_dir / "results-python-transmitter.txt"
        
    def log_transmission(self, stats):
        with open(self.log_file, 'a') as f:
            f.write("\n" + "="*50 + "\n")
            f.write(f"Transmission at {datetime.now().isoformat()}\n")
            f.write(f"File: {stats['file_name']}\n")
            f.write(f"File Size: {stats['file_size']/1024:.2f} KB\n")
            f.write(f"Duration: {stats['duration']:.2f} seconds\n")
            f.write(f"Throughput: {stats['throughput']/1024:.2f} KB/s\n")
            f.write(f"Total Packets: {stats['total_packets']}\n")
            f.write(f"ACKs Received: {stats['acks_received']}\n")
            f.write(f"Retries: {stats['retries']}\n")
            f.write(f"Success Rate: {stats['success_rate']:.1f}%\n")
            f.write("="*50 + "\n")

class Transmitter:
    transmission_id = 0  # Class variable to track transmission IDs
    DEFAULT_TIMEOUT = 1.0  # Default timeout in seconds

    def __init__(self, target_ip_address, target_port):
        self.udp_client_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.udp_client_socket.settimeout(self.DEFAULT_TIMEOUT)
        self.target_address = (target_ip_address, target_port)
        self.acks_received = 0
        self.retries = 0
        self.logger = TransmitterLogger()
        print("UDP socket created and bound.")

    def send_packet_with_ack(self, packet):
        """Send a packet and wait for ACK"""
        max_retries = 3
        retries = 0
        
        while retries < max_retries:
            try:
                # Send the packet
                packet_in_bytes = packet.serialization()
                self.udp_client_socket.sendto(packet_in_bytes, self.target_address)
                print(f"Sending packet tx_id={packet.transmission_id} seq={packet.sequence_number}")
                
                # Wait for ACK
                while True:
                    try:
                        data, addr = self.udp_client_socket.recvfrom(1024)
                        if data == b'ACK':
                            self.acks_received += 1
                            print(f"Received ACK for packet {packet.sequence_number}")
                            return True
                    except socket.timeout:
                        print(f"Timeout waiting for ACK for packet {packet.sequence_number}")
                        break
                
                retries += 1
                self.retries += 1
                print(f"Retrying packet {packet.sequence_number} (attempt {retries}/{max_retries})")
                
            except Exception as e:
                print(f"Error sending packet: {e}")
                retries += 1
        
        print(f"Failed to send packet {packet.sequence_number} after {max_retries} attempts")
        return False

    def send_packets(self, packets: list, file_name: str, file_size: int):
        """Send packets using Stop-and-Wait protocol"""
        total_packets = len(packets)
        start_time = time.time()
        
        for packet in packets:
            if not self.send_packet_with_ack(packet):
                print(f"Failed to send packet {packet.sequence_number}, aborting transmission")
                self.close_socket()
                return False
        
        end_time = time.time()
        duration = end_time - start_time
        throughput = file_size / duration
        
        stats = {
            'file_name': file_name,
            'file_size': file_size,
            'duration': duration,
            'throughput': throughput,
            'total_packets': total_packets,
            'acks_received': self.acks_received,
            'retries': self.retries,
            'success_rate': (self.acks_received/total_packets)*100
        }
        
        # Log statistics
        self.logger.log_transmission(stats)
        
        print(f"\nTransmission Statistics:")
        print(f"File: {file_name}")
        print(f"File Size: {file_size/1024:.2f} KB")
        print(f"Duration: {duration:.2f} seconds")
        print(f"Throughput: {throughput/1024:.2f} KB/s")
        print(f"Total packets sent: {total_packets}")
        print(f"Total ACKs received: {self.acks_received}")
        print(f"Total retries needed: {self.retries}")
        print(f"Success rate: {(self.acks_received/total_packets)*100:.1f}%")
        
        self.close_socket()
        return True

    def close_socket(self) -> str:
        self.udp_client_socket.close()
        return "Socket closed"

def run_transmission(file_name, target_ip, transmission_id):
    file_size = Path(file_name).stat().st_size
    builder = PacketBuilder(file_name, 1024, transmission_id)
    tx = Transmitter(target_ip, 4010)
    success = tx.send_packets(builder.get_all_packets(), file_name, file_size)
    
    if not success:
        print(f"Transmission {transmission_id} failed")
        return False
    else:
        print(f"Transmission {transmission_id} completed successfully")
        return True

def main():
    # Get file name from command line or use default
    file_name = sys.argv[1] if len(sys.argv) > 1 else "test_50mb.bin"
    target_ip = sys.argv[2] if len(sys.argv) > 2 else "127.0.0.1"
    
    print(f"\nStarting 10 transmissions of {file_name} to {target_ip}")
    print("="*50)
    
    successful_transmissions = 0
    for i in range(10):
        current_tx_id = Transmitter.transmission_id
        Transmitter.transmission_id += 1
        
        print(f"\nStarting transmission {i+1}/10 (ID: {current_tx_id})")
        if run_transmission(file_name, target_ip, current_tx_id):
            successful_transmissions += 1
        print(f"Completed transmission {i+1}/10")
        print("-"*50)
    
    print(f"\nTransmission Summary:")
    print(f"Total transmissions: 10")
    print(f"Successful: {successful_transmissions}")
    print(f"Failed: {10 - successful_transmissions}")
    print(f"Success rate: {(successful_transmissions/10)*100:.1f}%")
    print(f"Results have been saved to results/results-python-transmitter.txt")

if __name__ == "__main__":
    main()
